lib/my_enum.ex

```
@@ -223,4 +223,132 @@ defmodule MyEnum do
       false -> drop_while(t, function, result ++ [h])
     end
   end
+
+  @doc """
+    Enum.each/2 요소를 순회하며 주어진 함수에 요소값을 넣어 실행합니다.
+    요소를 모두 순회하면 :ok를 반환합니다.
+  """
+
+  def each([], _function), do: :ok
+
+  def each([h | t], function) do
+    function.(h)
+    each(t, function)
+  end
+
+  @doc """
+    Enum.empty?/1 리스트가 비어있는지 확인합니다.
+    비어있으면 true 그렇지 않으면 false를 반환합니다.
+  """
+
+  def empty?([]), do: true
+  def empty?([_h | _t]), do: false
+  def empty?(_not_list), do: :error
+
+  @doc """
+    fetch!/2 주어진 index의 요소를 반환합니다.
+    index는 zero base 입니다.
+    index가 음수로 전해지면 요소의 맨 뒤에서 부터 순회합니다.
+    index가 리스트의 범위를 벗어나면 OutOfBoundsError가 발생합니다.
+    해당 index에 요소가 없다면 :error를 반환합니다.
+  """
+  def fetch!([], _index), do: :error
```

함수명에 ! 느낌표가 붙는 함수는, "에러를 그냥 내겠다, Let it crash" 라는 뜻입니다.
보통 함수명에 느낌표가 붙지않으면, `{:ok, 성공값}`, `{:error, 에러값}` 으로 반환하고, 느낌표가 붙는다면 `성공값` 만 반환하고 에러는 그냥 크래쉬 시킵니다.
함수를 수정해서 예외는 그냥 let it crash 하세요.

---

그렇다면 함수명에 ? 물음표가 붙는 함수는 혹시 boolean값을 반환하는 함수일까요 ?

---

반드시는 아니지만 보통 그렇게해용